### PR TITLE
Initial attempt at supporting ESP-IDF 5.0.0

### PR DESCRIPTION
**Pull request by kbx81**
# What does this implement/fix?

This is just a batch of changes I found necessary to get a successful build with IDF 5.0.

It's...a mess...but we have to start somewhere. My hope is that others will find at least some of this work helpful as we move towards full, proper support of the updated framework.

In general, this is very much a WIP and I don't expect we'll merge this specific PR...but, who knows. 😄

Version 6.0 of the platform also seems to complain about `printf` (and siblings) all over the place. I'm guessing this is some compiler setting but I don't know how and didn't take the time to look up how to fix that...so for now I have a commit with that nonsense. You'll see it complain in any other component(s) where those functions are used.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

Note that, for now, this **requires** importing the mDNS component via @jesserockz's #4000 PR; but, with this, I was able to get a successful build. That said, mDNS **will not actually work** because we have more work to do to get the hostname set correctly. You'll probably notice warnings about this if/when you build.

That said, _if you run this on your ESP, you can still do updates OTA, but you'll need to use the device's IP as it won't respond to any hostname._

```yaml
# Example config.yaml
esphome:
  platformio_options:
    build_flags:
      - -DCONFIG_FREERTOS_ENABLE_BACKWARD_COMPATIBILITY

esp32:
  board: featheresp32
  framework:
    type: esp-idf
    version: "5.0.0"
    platform_version: "6.0.0"
    components:
      - name: esp-protocols
        source: github://espressif/esp-protocols

mdns:
  disabled: true
```

**NOTE:** I also ran into an issue building as described in [this thread](https://github.com/espressif/esp-idf/issues/8489); read the comments for the resolution (basically just run `install.sh` followed by `. export.sh`).

## Checklist:
  - [x] The code change is tested and works locally...mostly, kind of.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by oxan** [comment]
Some discussion on the discord: https://discord.com/channels/429907082951524364/1076282397903048725

**Comment by stintel** [comment]
Tested the changes I just pushed on ESP32-C3-DevKitM-1 and WEMOS LOLIN D32 PRO using the following version combinations:

* esp-idf 4.4.4 + platform-espressif 5.3.0
* esp-idf 5.0.1 + platform-espressif 6.0.0

The example config can be reduced to this:
```yaml
---
esphome:
  name: idf501

esp32:
  board: featheresp32
  framework:
    type: esp-idf
    version: "5.0.1"
    platform_version: "6.0.0"

```

FYI: platform-espressif32 >= 6.0.1 doesn't work due to https://github.com/platformio/platform-espressif32/commit/8507dd8e7fc6d3fb56ec5372fc24a3044226a058. It now requires python-venv. I could fix that, but might be getting out of scope for this PR. Please let me know what people prefer.

File: esphome/components/mdns/__init__.py [hunk 1]
```diff
@@ -79,6 +84,14 @@ async def to_code(config):
         elif CORE.is_rp2040:
             cg.add_library("LEAmDNS", None)
 
+    if CORE.using_esp_idf and (framework_ver.major, framework_ver.minor) >= (5, 0):
```
Comment (oxan) [inline esphome/components/mdns/__init__.py]: Minor nit; but you can compare `cv.Version` instances directly:
```suggestion
    if CORE.using_esp_idf and framework_ver >= cv.Version(5, 0, 0):
```
(or maybe even drop the `framework_ver` variable).

File: esphome/core/config.py [hunk 2]
```diff
@@ -383,6 +386,9 @@ async def to_code(config):
     if CORE.using_arduino:
         CORE.add_job(add_arduino_global_workaround)
 
+    if CORE.using_esp_idf and (framework_ver.major, framework_ver.minor) >= (5, 0):
```
Comment (oxan) [inline esphome/core/config.py]: same here.

**Review by oxan — APPROVED** [review]
LGTM.

**Comment by stintel** [comment]
Retested the changes on ESP32-C3-DevKitM-1 and WEMOS LOLIN D32 PRO using the following version combinations:

    esp-idf 4.4.4 + platform-espressif 5.3.0
    esp-idf 5.0.1 + platform-espressif 6.0.0

This time also with IPv6 enabled and disabled, due to the changes I made to resolve conflicts after d42f35de5d54. @jesserockz can you please verify I've correctly resolved these conflicts?

I also noticed that switching esp-idf versions requires wiping esp-idf's python dependencies installed by platformio. This is exactly what they fixed in https://github.com/platformio/platform-espressif32/commit/8507dd8e7fc6d3fb56ec5372fc24a3044226a058, so I've added a commit to install python3-venv.

Afterwards retested with platform-espressif32 6.0.1 and 6.1.0.

New example config to avoid those python dependency issues:
```yaml
---
esphome:
  name: idf501

esp32:
  board: featheresp32
  framework:
    type: esp-idf
    version: "5.0.1"
    platform_version: "6.1.0"
```

**Review by jesserockz — APPROVED** [review]
Changes look fine. Will give them a quick test locally during the week if not today.

File: esphome/core/config.py [hunk 3]
```diff
@@ -383,6 +384,11 @@ async def to_code(config):
     if CORE.using_arduino:
         CORE.add_job(add_arduino_global_workaround)
 
+    if CORE.using_esp_idf and CORE.data[KEY_CORE][KEY_FRAMEWORK_VERSION] >= cv.Version(
+        5, 0, 0
+    ):
+        cg.add_build_flag("-Wno-format")
```
Comment (jesserockz) [inline esphome/core/config.py]: Is this still required after https://github.com/esphome/esphome/pull/4671 ?

Comment (stintel) [reply esphome/core/config.py]: No. Dropped the commit.

**Comment by stintel** [comment]
Anything still blocking this PR from being merged?

**Review by jesserockz — APPROVED** [review]

**Pull request merged**